clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java

```diff
@@ -0,0 +1,132 @@
+ * and equals {@code consumer}, the {@link AsyncKafkaConsumer} will be returned. Otherwise, the
+ * {@link LegacyKafkaConsumer} will be returned.
+ *
+ * <p/>
+ *
+ * This is not to be called by end users and callers should not attempt to determine the underlying implementation
+ * as this will make such code very brittle. Users of this facility should honor the top-level {@link Consumer} API
+ * contract as-is.
+ */
+public class ConsumerDelegateCreator {
+
+    /**
+     * This is it! This is the core logic. It's extremely rudimentary.
+     */
+    private static boolean useNewConsumer(Map<?, ?> configs) {
+        Object groupProtocol = configs.get("group.protocol");
+
+        // Takes care of both the null and type checks.
+        if (!(groupProtocol instanceof String))
+            return false;
+
+        return ((String) groupProtocol).equalsIgnoreCase("consumer");
+    }
+
+    public <K, V> Consumer<K, V> create(Map<String, Object> configs) {
+        return createInternal(() -> {
+            if (useNewConsumer(configs)) {
+                return new AsyncKafkaConsumer<>(configs);
+            } else {
+                return new LegacyKafkaConsumer<>(configs);
+            }
+        });
+    }
+
+    public <K, V> Consumer<K, V> create(Properties properties) {
+        return createInternal(() -> {
+            if (useNewConsumer(properties)) {
+                return new AsyncKafkaConsumer<>(properties);
+            } else {
+                return new LegacyKafkaConsumer<>(properties);
+            }
+        });
+    }
+
+    public <K, V> Consumer<K, V> create(Properties properties,
+                                        Deserializer<K> keyDeserializer,
+                                        Deserializer<V> valueDeserializer) {
+        return createInternal(() -> {
+            if (useNewConsumer(properties)) {
+                return new AsyncKafkaConsumer<>(properties, keyDeserializer, valueDeserializer);
+            } else {
+                return new LegacyKafkaConsumer<>(properties, keyDeserializer, valueDeserializer);
+            }
+        });
+    }
+
+    public <K, V> Consumer<K, V> create(Map<String, Object> configs,
+                                        Deserializer<K> keyDeserializer,
+                                        Deserializer<V> valueDeserializer) {
+        return createInternal(() -> {
+            if (useNewConsumer(configs)) {
+                return new AsyncKafkaConsumer<>(configs, keyDeserializer, valueDeserializer);
+            } else {
+                return new LegacyKafkaConsumer<>(configs, keyDeserializer, valueDeserializer);
+            }
+        });
+    }
+
+    public <K, V> Consumer<K, V> create(ConsumerConfig config,
+                                        Deserializer<K> keyDeserializer,
+                                        Deserializer<V> valueDeserializer) {
+        return createInternal(() -> {
+            if (useNewConsumer(config.values())) {
```

Comment:
There's a lot of duplicated code for
```
if (useNewConsumer(config.values())) {
                return new AsyncKafkaConsumer<>(config, keyDeserializer, valueDeserializer);
            } else {
                return new LegacyKafkaConsumer<>(config, keyDeserializer, valueDeserializer);
            }
```

Reply:
Yep. I don't really care for it, either. As above, I wanted to preserve a 1:1 mapping between the constructor that was invoked in `KafkaConsumer` and the constructor in the delegate. If we feel that's overkill, we can change it.

Reply:
I think that we should simplify the constructors in the delegates. It does not make sense to keep all of them.

Reply:
I've refactored the code to use fewer constructors in the delegates and pass a fully-constructed `ConsumerConfig` to the delegates.